### PR TITLE
Improve efficiency of tapering hamiltonians

### PR DESCRIPTION
**Pull request by Jaybsoni**
**Context:**
The method for Tapering hamiltonians requires performing repeated multiplication between hamiltonians. This was previously done using the `_observable_mult` method. This is the computational bottle neck of the procedure and needs to be improved. 

**Description of the Change:**
Use pauli arithmetic to perform the multiplication more efficiently.

### CODE REVIEW

File: tests/pauli/test_pauli_arithmetic.py [hunk 1]
```diff
@@ -565,14 +565,6 @@ def test_hamiltonian_wire_order(self):
 
         assert qml.equal(op, id)
 
-    def test_hamiltonian_complex_coeffs(self):
```
Comment (soranjh) [inline tests/pauli/test_pauli_arithmetic.py]: Why this test is removed?

File: pennylane/qchem/tapering.py [hunk 2]
```diff
@@ -216,40 +216,6 @@ def paulix_ops(generators, num_qubits):
     return paulixops
 
 
-def _observable_mult(obs_a, obs_b):
```
Comment (soranjh) [inline pennylane/qchem/tapering.py:219]: Should we keep this (maybe deprecate) for some time before deleting it?

**Review by soranjh — COMMENTED** [review]
Thanks @Jaybsoni, it might be very nice if you have a plot comparing the times for the `taper` and `taper_hf` functions, for a qubit range,  before and after the changes.

File: tests/pauli/test_pauli_arithmetic.py (hunk 1 above)
Comment (Jaybsoni) [reply tests/pauli/test_pauli_arithmetic.py]: It is deleted in both the base branch and here. Addressing an old edge case I fixed.

File: pennylane/qchem/tapering.py (hunk 2 above)
Comment (Jaybsoni) [reply pennylane/qchem/tapering.py:219]: We could. I assumed that it's a private method so only we should be the ones using it. I think its fine to remove out-right

**Comment by Jaybsoni** [comment]
[sc-34786]

**Comment by obliviateandsurrender** [comment]
I ran some benchmarks from my side, improvements look good! 
image

**Review by soranjh — APPROVED** [review]
Looks good, thanks @Jaybsoni.

**Review by obliviateandsurrender — APPROVED** [review]
